src/serde_support.rs

```diff
@@ -0,0 +1,275 @@
+        }
+        seq.end()
+    }
+}
+
+impl<T> Serialize for CircularQueue<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let to_ser = CircularQueueSerialize {
+            length: self.len(),
+            values: self.asc_iter(),
+        };
+
+        let mut seq = serializer.serialize_struct("CircularQueue", 2)?;
+        seq.serialize_field("capacity", &self.capacity)?;
+        seq.serialize_field("values", &to_ser)?;
+        seq.end()
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(field_identifier, rename_all = "lowercase")]
+enum Field {
+    Capacity,
+    Values,
+}
+
+#[derive(Debug)]
+/// `serde::de::Visitor` for a circular queue
+struct CircularQueueVisitor<T> {
+    marker: PhantomData<fn() -> CircularQueue<T>>,
+}
+
+impl<T> CircularQueueVisitor<T> {
+    pub fn new() -> CircularQueueVisitor<T> {
+        CircularQueueVisitor {
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<'de, T> Visitor<'de> for CircularQueueVisitor<T>
+where
+    T: Deserialize<'de>,
+{
+    type Value = CircularQueue<T>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "a circular queue")
+    }
+
+    fn visit_seq<S>(self, mut seq: S) -> Result<Self::Value, S::Error>
+    where
+        S: SeqAccess<'de>,
+    {
+        let capacity = match seq.next_element()? {
+            Some(val) => val,
+            None => return Err(Error::missing_field("capacity")),
+        };
+        let values = match seq.next_element::<Vec<T>>()? {
+            Some(val) => val,
+            None => return Err(Error::missing_field("values")),
+        };
+        let mut ret = CircularQueue::with_capacity(capacity);
+        for x in values {
+            ret.push(x);
+        }
+        Ok(ret)
+    }
+
+    fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
+    where
+        V: MapAccess<'de>,
+    {
+        let mut capacity = None;
+        let mut values = None;
+        while let Some(key) = map.next_key()? {
+            match key {
+                Field::Capacity => {
+                    if capacity.is_some() {
+                        return Err(Error::duplicate_field("capacity"));
+                    }
+                    capacity = Some(map.next_value()?);
+                }
+                Field::Values => {
+                    if values.is_some() {
+                        return Err(Error::duplicate_field("values"));
+                    }
+                    values = Some(map.next_value::<Vec<T>>()?);
+                }
+            }
+        }
+        let capacity = capacity.ok_or_else(|| Error::missing_field("capacity"))?;
+        let values = values.ok_or_else(|| Error::missing_field("values"))?;
+        let mut ret = CircularQueue::with_capacity(capacity);
+        for x in values {
+            ret.push(x);
+        }
```

Comment:
I wonder if this could be implemented automatically by having
```rust
#[derive(Deserialize)]
struct CircularQueueStruct<T> {
    capacity: usize,
    values: Vec<T>,
}
```
and then deserializing that in `fn deserialize` and just moving over the values like is done in the end of this function? If this works, it should remove the need for all this manual struct deserialization.

Reply:
Yes.  I struggled a bit with how exactly the `deserialize` function was supposed to be invoked, but then it clicked and made sense.  Moving the vector requires handling the case that the serialized vector is too big for the capacity, so I guard against that.